composer/trainer/trainer.py

```diff
@@ -1091,8 +1081,94 @@ def _train_loop(self) -> None:
+        else:
+            self.state.grad_accum = min(2 * self.state.grad_accum, self.state.batch_num_samples)
+            self.logger.data_batch({'trainer/grad_accum': self.state.grad_accum})
+
+    def _compute_metrics(self, train_metrics: Union[MetricCollection, None]):
+        """Compute training metrics.
+
+        Args:
+            train_metrics (Union[MetricCollection, None]): Existing train metrics
+        """
+        # Compute metrics on the training set
+        if self._compute_training_metrics:
+            assert train_metrics is not None
+            self.state.model.eval()
+            with torch.no_grad():
+                for eval_microbatch in self._train_data_spec.split_batch(self.state.batch, self.state.grad_accum):
+                    # TODO: Detect if self.run_event(Event.AFTER_DATALOADER) changes the training
+                    # data and if so print a warning that metrics may return unexpected results
+                    outputs, targets = self._original_model.validate(eval_microbatch)
+                    train_metrics.update(outputs, targets)
+
+    def _train_batch(self, use_grad_scaling: bool):
+        """Compute loss by training on a full batch of data. Adaptively change microbatch size if enabled to maximize
+        GPU usage.
+
+        Args:
+            use_grad_scaling (bool): Enables gradient scaling
+        """
+        # Retry until we successfully complete training and return loss
+        while True:
+            total_loss = None
+            should_handle_cuda_oom = False
+            caught_error = None
+            try:
+                assert self.state.scaler is not None
+                microbatches = self._train_data_spec.split_batch(self.state.batch, self.state.grad_accum)
+                if self.deepspeed_enabled:
+                    total_loss = self._train_microbatches(microbatches)
+                elif self._use_closures():
+                    for optimizer in self.state.optimizers:
+                        if use_grad_scaling:
+                            total_loss = self.state.scaler.step(
+                                optimizer, closure=lambda **kwargs: self._train_microbatches(microbatches, **kwargs))
+                        else:
+                            total_loss = optimizer.step(
+                                closure=lambda **kwargs: self._train_microbatches(microbatches, **kwargs).item())
+                else:
+                    total_loss = self._train_microbatches(microbatches)
+                    for optimizer in self.state.optimizers:
+                        if use_grad_scaling:
+                            self.state.scaler.step(optimizer)
+                        else:
+                            optimizer.step()
+            except RuntimeError as e:
+                if self._is_cuda_oom(e):
+                    should_handle_cuda_oom = True
+                else:
+                    caught_error = e
```

Comment:
Here's the scenario I'm thinking about: 
1. rank 0 cuda OOMs. `should_handle_cuda_oom` is set to True
2. the sync strategy is set to sync every microbatch. Rank 1 does not crash and waits for 0 at end of microbatch. Eventually, it times out (since 0 never reaches there) and throws an error)
3. Rank 1 has `catch_error` set to be `e`, which is something related to a timeout at a barrier. It ignores this since someone else got an OOM (the root cause of the issue) and everyone reruns with higher `grad_accum`

Reply:
Ah, I see. In this case, can we explicitly catch the timeout errors?

Reply:
yes, very good point, will definitely add a few comments to make this more clear because its subtle